draft-ietf-tls-tls13.md

```diff
@@ -2810,32 +2827,82 @@ configuration identifier for each connection, thus preventing
+is advisable.  However, any allowance also increases the opportunity
+for replay.  In this case, it is better to reject early data than to
+risk greater exposure to replay attacks.
+
+#### Ticket Age
+
+%%% Key Exchange Messages
+       struct {
+           uint64 ticket_age;
+       } TicketAge;
+
+When the client sends the "early_data" extension, it MUST also send
+a "ticket_age" extension in its EncryptedExtensions block. This value
+contains the time elapsed since the client learned about the server
+configuration that it is using, in milliseconds.  This value can
+be used by the server to limit the time over which early data can
+be replayed.
+       
+
 ### Server Parameters
 
 ####  Encrypted Extensions
 
 When this message will be sent:
 
-> The EncryptedExtensions message MUST be sent immediately after the
+> In all handshakes, the server MUST send the
+EncryptedExtensions message MUST be sent immediately after the
 ServerHello message. This is the first message that is encrypted
-under keys derived from ES.
+under keys derived from ES. If the client indicates "early_data"
+in its ClientHello, it MUST also send EncryptedExtensions immediately
+following the ClientHello and immediately prior to the Finished.
 
 Meaning of this message:
 
-> The EncryptedExtensions message simply contains any extensions
+> The EncryptedExtensions message contains any extensions
 which should be protected, i.e., any which are not needed to
-establish the cryptographic context. The same extension types
-MUST NOT appear in both the ServerHello and EncryptedExtensions.
-If the same extension appears in both locations, the client
-MUST rely only on the value in the EncryptedExtensions block.
-All server-sent extensions other than those explicitly
-listed in {{server-hello}} or designated in the IANA registry
-MUST only appear in EncryptedExtensions. Extensions which
-are designated to appear in ServerHello MUST NOT appear in
-EncryptedExtensions. Clients MUST check EncryptedExtensions
-for the presence of any forbidden extensions and if
-any are found MUST terminate the handshake with
-an "illegal_parameter" alert.
+establish the cryptographic context.
+
+The same extension types MUST NOT appear in both the ServerHello and
+EncryptedExtensions.  If the same extension appears in both locations,
+the client MUST rely only on the value in the EncryptedExtensions
+block.  All server-sent extensions other than those explicitly listed
+in {{server-hello}} or designated in the IANA registry MUST only
+appear in EncryptedExtensions. Extensions which are designated to
+appear in ServerHello MUST NOT appear in EncryptedExtensions. Clients
+MUST check EncryptedExtensions for the presence of any forbidden
+extensions and if any are found MUST terminate the handshake with an
+"illegal_parameter" alert.
+
+The client's EncryptedExtensions apply only to the early data
+with which they appear. Servers MUST NOT use them to negotiate
+the rest of the handshake.
```

Comment:
Maybe add, "Only those extensions explicitly designated as being included in 0-RTT Encrypted Extensions in the IANA registry can be sent in the client's EncryptedExtensions."

I think that the check  above can be made mandatory for both client and server.  I have no problem tearing a connection down if 0-RTT includes bad extensions, though rejecting 0-RTT might be less disruptive.